### PR TITLE
sign and recover tool

### PR DESCRIPTION
**Pull request by Tigo9527**
This change is [Reviewable](https://reviewable.io/reviews/conflux-chain/web3pay-service/2)

### CODE REVIEW

File: cmd/ecrecover.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	decode, err := hexutil.Decode(signature)
+	decode[64] -= 27
+	hash := crypto.Keccak256Hash([]byte(BuildPersonalSignMessage(message)))
+	sigPublicKey, err := crypto.SigToPub(hash.Bytes(), decode)
+	if err != nil {
+		return "", err
+	}
+	address := crypto.PubkeyToAddress(*sigPublicKey).Hex()
+	return address, nil
+}
```
Comment (wanliqun) [inline cmd/ecrecover.go]: Seems these functions can be reused for RESTful API, better put them within some common or utility package?

File: cmd/ecrecover.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
+/*
+Copyright © 2022 NAME HERE <EMAIL ADDRESS>
+
+*/
```
Comment (wanliqun) [inline cmd/ecrecover.go]: Could be removed since they are nonsense.

File: cmd/ecrecover.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+// Returns a signature string
+func PersonalSign(message string, privateKey *ecdsa.PrivateKey) (string, common.Hash, error) {
+	fullMessage := BuildPersonalSignMessage(message)
+	hash := crypto.Keccak256Hash([]byte(fullMessage))
+	signatureBytes, err := crypto.Sign(hash.Bytes(), privateKey)
+	if err != nil {
+		return "", hash, err
+	}
+	signatureBytes[64] += 27
```
Comment (wanliqun) [inline cmd/ecrecover.go]: Can you add some comment to explain what for to add 27?

File: cmd/ecrecover.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	if err != nil {
+		return "", err
+	}
+	address := crypto.PubkeyToAddress(*sigPublicKey).Hex()
+	return address, nil
+}
+
+// ecrecoverCmd represents the ecrecover command
+var ecrecoverCmd = &cobra.Command{
+	Use:        "recover signature message",
```
Comment (wanliqun) [inline cmd/ecrecover.go]: too long as a command

File: cmd/ecrecover.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	Short:      "Recover address",
+	Long:       `Recover address from message and its signature`,
+	Args:       cobra.ExactArgs(2),
+	ArgAliases: []string{"signature", "message"},
+	Run: func(cmd *cobra.Command, args []string) {
+		fmt.Println("ecrecover called", args)
```
Comment (wanliqun) [inline cmd/ecrecover.go]: Can use logrus so we can have fine grained control over logging output

File: cmd/ecrecover.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+		address, err := RecoverAddress(message, signature)
+		if err != nil {
+			log.Fatal("recover fail", err)
+		}
+		fmt.Printf("recovered address: %s\n", address)
+	},
+}
+
+var signCmd = &cobra.Command{
+	Use:        "sign privateKey message",
```
Comment (wanliqun) [inline cmd/ecrecover.go]: ditto

File: cmd/ecrecover.go (hunk 1 above)
Comment (Tigo9527) [reply cmd/ecrecover.go]: Moved to util package.

File: cmd/ecrecover.go (hunk 2 above)
Comment (Tigo9527) [reply cmd/ecrecover.go]: Removed.

File: cmd/ecrecover.go (hunk 3 above)
Comment (Tigo9527) [reply cmd/ecrecover.go]: Comment added.

File: cmd/ecrecover.go (hunk 4 above)
Comment (Tigo9527) [reply cmd/ecrecover.go]: Added <>

File: cmd/ecrecover.go (hunk 5 above)
Comment (Tigo9527) [reply cmd/ecrecover.go]: Used

File: cmd/ecrecover.go [hunk 7]
```diff
@@ -44,15 +44,15 @@ var signCmd = &cobra.Command{
...
+		logrus.Info("sign with address ", addressSig)
 
 		input := message
 		signature, _, err := util.PersonalSign(input, privateKey)
 		if err != nil {
 			log.Fatal(err)
 		}
 
-		logrus.Info("signature", signature, "length", len(signature))
+		logrus.Info("signature ", signature, "length ", len(signature))
```
Comment (wanliqun) [inline cmd/ecrecover.go]: The  format might be ugly, the recommended way to use is:
`logrus.WithFields(...).Info(...)`

**Review by wanliqun — APPROVED** [review]

**Pull request merged**